File: simulation/simulation.py
from . import config
import random
from .organism import Organism
from .world import World

#Initialize organism lists
organisms = []
dead_organisms = []

#Create world
world = World(config.WIDTH, config.HEIGHT, config.FOOD_NUMBER)

#Save occupied positions
occupied_positions = set()

def build_organism(x, y, energy):
    trait_name = config.TRAIT_NAME
    mutable_val = config.STARTING_MUTABLE_VALUE
    speed = mutable_val if trait_name == "speed" else 1
    efficiency = mutable_val if trait_name == "efficiency" else 1.0
    return Organism(x, y, speed=speed, energy=energy, efficiency=efficiency)

def initialize_organisms():
    """Initialize organisms on the grid in random locations"""
    global organisms, dead_organisms, occupied_positions, world
    organisms.clear()
    dead_organisms.clear()
    occupied_positions.clear()
    
    for _ in range(config.NUM_ORGANISMS):
        while True:
            x = random.randint(0, config.WIDTH - 1)
            y = random.randint(0, config.HEIGHT - 1)
            #Check if position occupied, if not place organism
            if (x, y) not in occupied_positions:
                organisms.append(build_organism(x, y, config.STARTING_ENERGY))
                world.place_organism(x, y)
                occupied_positions.add((x, y))
                break

def reset_simulation():
    """Reset the simulation to initial state"""
    global world, organisms, dead_organisms, occupied_positions
    world = World(config.WIDTH, config.HEIGHT, config.FOOD_NUMBER)
    initialize_organisms()

#Initialize organisms on the grid in random locations
initialize_organisms()

#Get the distribution of the tested trait
def get_trait_distribution(trait_name, possible_values):
    counts = {val: 0 for val in possible_values}
    total = len(organisms)

    for org in organisms:
        value = getattr(org, trait_name)
        if value in counts:
            counts[value] += 1

    return [counts[val] / total if total > 0 else 0 for val in possible_values]


#Run it!
def simulate():
    global organisms, dead_organisms, world

    #End if all organisms die
    if not organisms:
        return False

    # Clear the world grid before processing organisms
    for y in range(world.height):
        for x in range(world.width):
            world.grid[y][x] = None

    live_organisms = []
    new_offspring = []  # Collect offspring separately to avoid modifying list while iterating

    #All organisms will...
    for org in organisms:
        
        #Move and eat
        print(f"Org at ({org.x}, {org.y}) Before move: energy={org.energy}, speed={org.speed}, efficiency={org.efficiency}")
        org.move(world)
        print(f"Org at ({org.x}, {org.y}) After move: energy={org.energy}")

        #Or die.
        if org.check_survival(world):
            # Check if position is empty before placing (should be after move)
            if world.is_empty(org.x, org.y):
                world.place_organism(org.x, org.y)
            else:
                print(f"WARNING: Organism at ({org.x}, {org.y}) tried to occupy non-empty cell!")

            # Handle reproduction
            offspring = org.reproduce(world)
            if offspring:
                new_offspring.append(offspring)
                if world.is_empty(offspring.x, offspring.y):
                    world.place_organism(offspring.x, offspring.y)
                    print(f"New offspring born at ({offspring.x}, {offspring.y}) with speed {offspring.speed}")
                else:
                    print(f"WARNING: Offspring at ({offspring.x}, {offspring.y}) tried to occupy non-empty cell!")

            live_organisms.append(org)
        else:
            # Organism died - clear its position but keep it for death animation
            trait_value = getattr(org, config.TRAIT_NAME)
            world.clear(org.x, org.y)
            dead_organisms.append({
                'x': org.x, 
                'y': org.y, 
                'frames_left': config.DEATH_ANIMATION_FRAMES,
                'trait': int(trait_value)  # Store trait for death animation
            })
            print(f"Organism died at ({org.x}, {org.y})")

    # Add new offspring to the organisms list (modify in-place to update global)
    organisms.clear()
    organisms.extend(live_organisms)
    organisms.extend(new_offspring)
    
    # Debug: verify speed distribution after adding offspring
    if new_offspring:
        print(f"Added {len(new_offspring)} new offspring. Total organisms now: {len(organisms)}")
        for off in new_offspring:
            print(f"  - Offspring at ({off.x}, {off.y}) with speed {off.speed}")
    
    # Debug: final speed distribution
    final_speeds = {}
    for org in organisms:
        speed = getattr(org, config.TRAIT_NAME)
        final_speeds[speed] = final_speeds.get(speed, 0) + 1
    print(f"Final organisms by speed in simulate(): {final_speeds}")

    #Update dead organism animations
    for corpse in dead_organisms:
        corpse['frames_left'] -= 1
    dead_organisms[:] = [corpse for corpse in dead_organisms if corpse['frames_left'] > 0]

    #Replenish the food sources
    world.replenish_food()

    #Keep going if we have some live organisms
    return True

#Driver when running simulation file
def run():
    for step in range(config.SIMULATION_STEPS):
        if not simulate():
            print(f"Simulation has ended at step {step}.")
            break
        print(f"Step {step}: ")
        for row in world.grid:
            print(row)

if __name__ == "__main__":
    run()

File: src/sim/simulation.js
// The Simulation orchestrates one full run: it owns the World, the live
// `organisms` list, and a `deadOrganisms` list used for the death animation.
//
// One important difference vs. the Python original: that file kept everything
// at module scope (globals + import-time `initialize_organisms()`). Here it's
// a class so we can throw away the whole instance and build a fresh one
// whenever the user changes settings or hits Reset.

import { defaultConfig, TRAIT_POSSIBLE_VALUES, labelsFor } from './config.js'
import { World } from './world.js'
import { Organism } from './organism.js'
import { randInt } from './rand.js'

export class Simulation {
  constructor(userConfig = {}) {
    // Merge user overrides on top of the defaults. The merged object is
    // shared with World/Organism instances so they all see the same values.
    this.config = { ...defaultConfig, ...userConfig }

    this.world = new World(
      this.config.WIDTH,
      this.config.HEIGHT,
      this.config.FOOD_NUMBER,
      this.config,
    )

    this.organisms = []
    this.deadOrganisms = []

    this.initializeOrganisms()
  }

  // Spawn NUM_ORGANISMS organisms in unique random cells.
  initializeOrganisms() {
    const c = this.config
    const occupied = new Set()
    const key = (x, y) => `${x},${y}`

    for (let i = 0; i < c.NUM_ORGANISMS; i++) {
      // Keep rolling until we land on an empty cell.
      while (true) {
        const x = randInt(0, c.WIDTH - 1)
        const y = randInt(0, c.HEIGHT - 1)
        if (!occupied.has(key(x, y))) {
          this.organisms.push(this.buildOrganism(x, y, c.STARTING_ENERGY))
          this.world.placeOrganism(x, y)
          occupied.add(key(x, y))
          break
        }
      }
    }
  }

  // Whichever trait is currently being selected for starts at the configured
  // mutable value; the other traits are pinned at their neutral baseline.
  buildOrganism(x, y, energy) {
    const c = this.config
    const mutable = c.STARTING_MUTABLE_VALUE
    return new Organism({
      x, y, energy,
      speed:      c.TRAIT_NAME === 'speed'      ? mutable : 1,
      efficiency: c.TRAIT_NAME === 'efficiency' ? mutable : 1.0,
      config: c,
    })
  }

  reset() {
    // Build a brand new world and population. Cheaper and clearer than
    // surgically wiping every grid cell.
    this.world = new World(
      this.config.WIDTH,
      this.config.HEIGHT,
      this.config.FOOD_NUMBER,
      this.config,
    )
    this.organisms = []
    this.deadOrganisms = []
    this.initializeOrganisms()
  }

  // Run one tick of the simulation. Returns false if everyone is dead.
  step() {
    if (this.organisms.length === 0) return false

    // Wipe the organism layer of the grid before reprocessing. Each organism
    // re-places itself if it survives its move.
    for (let y = 0; y < this.world.height; y++) {
      for (let x = 0; x < this.world.width; x++) {
        this.world.grid[y][x] = null
      }
    }

    const liveOrganisms = []
    const newOffspring  = []

    for (const org of this.organisms) {
      org.move(this.world)

      if (org.checkSurvival(this.world)) {
        // Survived — claim our cell back and try to reproduce.
        if (this.world.isEmpty(org.x, org.y)) this.world.placeOrganism(org.x, org.y)

        const child = org.reproduce(this.world)
        if (child) {
          newOffspring.push(child)
          if (this.world.isEmpty(child.x, child.y)) {
            this.world.placeOrganism(child.x, child.y)
          }
        }

        liveOrganisms.push(org)
      } else {
        // Died — record a corpse for the death-fade animation.
        const traitValue = org[this.config.TRAIT_NAME]
        this.world.clear(org.x, org.y)
        this.deadOrganisms.push({
          x: org.x,
          y: org.y,
          framesLeft: this.config.DEATH_ANIMATION_FRAMES,
          trait: Math.trunc(traitValue),
        })
      }
    }

    this.organisms = liveOrganisms.concat(newOffspring)

    // Tick down corpse animations and prune the expired ones.
    for (const corpse of this.deadOrganisms) corpse.framesLeft -= 1
    this.deadOrganisms = this.deadOrganisms.filter((c) => c.framesLeft > 0)

    this.world.replenishFood()
    return true
  }

  // Compute proportions of each possible trait value among living organisms.
  // Returns an array aligned with TRAIT_POSSIBLE_VALUES[traitName].
  getTraitDistribution() {
    const traitName = this.config.TRAIT_NAME
    const possible = TRAIT_POSSIBLE_VALUES[traitName] || []
    const counts = Object.fromEntries(possible.map((v) => [v, 0]))
    const total = this.organisms.length

    for (const org of this.organisms) {
      const v = org[traitName]
      if (v in counts) counts[v]++
    }

    return possible.map((v) => (total > 0 ? counts[v] / total : 0))
  }

  // Snapshot of the current state, shaped exactly like the old Flask /state
  // response so the React components don't need to change anything.
  getState() {
    const c = this.config
    const w = this.world

    // Organism layer: '' for empty, trait value as string otherwise.
    const grid = []
    for (let y = 0; y < c.HEIGHT; y++) {
      const row = new Array(c.WIDTH).fill('')
      grid.push(row)
    }

    // Food layer: round to 1 decimal so React keys/renders are stable.
    const food = []
    for (let y = 0; y < c.HEIGHT; y++) {
      const row = new Array(c.WIDTH)
      for (let x = 0; x < c.WIDTH; x++) {
        row[x] = Math.round(w.food[y][x] * 10) / 10
      }
      food.push(row)
    }

    for (const org of this.organisms) {
      if (org.x >= 0 && org.x < c.WIDTH && org.y >= 0 && org.y < c.HEIGHT) {
        const traitInt = Math.trunc(org[c.TRAIT_NAME])
        grid[org.y][org.x] = String(traitInt)
      }
    }

    // Dead layer: keyed by "x,y" with progress 0..1 (0 = just died, 1 = gone).
    const dead = {}
    for (const corpse of this.deadOrganisms) {
      const { x, y, framesLeft, trait } = corpse
      if (x >= 0 && x < c.WIDTH && y >= 0 && y < c.HEIGHT) {
        const progress = 1 - framesLeft / c.DEATH_ANIMATION_FRAMES
        dead[`${x},${y}`] = { progress, trait }
      }
    }

    return {
      grid,
      food,
      alive: this.organisms.length > 0,
      trait_distribution: this.getTraitDistribution(),
      trait_name: c.TRAIT_NAME,
      trait_labels: labelsFor(c.TRAIT_NAME),
      dead,
    }
  }
}
